src/main/java/com/palantir/refreshable/DefaultRefreshable.java

```diff
@@ -100,64 +106,81 @@ public T current() {
     }
 
     @Override
-    public synchronized Disposable subscribe(Consumer<? super T> throwingSubscriber) {
-        SideEffectSubscriber<? super T> trackedSubscriber =
-                rootSubscriberTracker.newSideEffectSubscriber(throwingSubscriber, this);
-
-        Disposable disposable = subscribeToSelf(trackedSubscriber);
-        Disposable unsubscribeAndUntrack = () -> {
-            disposable.dispose();
-            rootSubscriberTracker.deleteReferenceTo(trackedSubscriber);
-        };
-        return unsubscribeAndUntrack;
+    public Disposable subscribe(Consumer<? super T> throwingSubscriber) {
+        readLock.lock();
+        try {
+            SideEffectSubscriber<? super T> trackedSubscriber =
+                    rootSubscriberTracker.newSideEffectSubscriber(throwingSubscriber, this);
+
+            Disposable disposable = subscribeToSelf(trackedSubscriber);
+            Disposable unsubscribeAndUntrack = () -> {
+                disposable.dispose();
+                rootSubscriberTracker.deleteReferenceTo(trackedSubscriber);
+            };
+            return unsubscribeAndUntrack;
+        } finally {
+            readLock.unlock();
+        }
     }
 
-    private synchronized Disposable subscribeToSelf(Consumer<? super T> subscriber) {
-        orderedSubscribers.add(subscriber);
-        int subscribers = orderedSubscribers.size();
-        if (subscribers > WARN_THRESHOLD) {
-            log.warn(
-                    "Refreshable {} has an excessive number of subscribers: {} and is likely leaking memory. "
-                            + "The current warning threshold is {}.",
-                    SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
-                    SafeArg.of("numSubscribers", subscribers),
-                    SafeArg.of("warningThreshold", WARN_THRESHOLD),
-                    new SafeRuntimeException("location"));
-        } else if (log.isDebugEnabled()) {
-            log.debug(
-                    "Added a subscription to refreshable {} resulting in {} subscriptions",
-                    SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
-                    SafeArg.of("numSubscribers", subscribers));
+    private void preSubscribeLogging() {
```

Comment:
@GuardedBy

Reply:
I don’t think this needs to be protected because the collection is already synchronized. At this point we could probably get rid of the logging because the benchmarks aren’t too bad even when refreshable isn’t used as originally intended, but IMO it’s still not a great idea to map in hot paths.

Reply:
Curious what you think we should do here. Keep or remove the logging?

if this code can run faster, it may also log a lot more in these scenarios as well. The  biggest problem in that case becomes log volume...

Reply:
Unfortunately the project I added the logging for only upgraded refreshables this week... unsure if it has even merged yet :-/

Reply:
The logging has detected quite a few things at this point. If this PR goes through it's less clear how important that will be, we may be able to churn away `map` calls without causing the system to grind to a halt, but we might as well leave the logging around to validate.